### ISSUE TITLE
Suggestions for making "variables" useful

### ISSUE THREAD

**Issue by dcjkfgdjhd** [cgeo/cgeo#13839]
### Tell us your idea!

Currently the "variables" tab is not very useful. 

It is very un-structured, it requires a lot of left and right swiping between the "description" and the "variables" tab, and it is very hard to know which formula in the variables tab belongs to which stage, or even which part of the next coordinates.

### What solution would you suggest?

I would completely get rid of the "variables" tab, and instead, when scanning the cache listing, turn every single detected formula into a clickable link.

When clicking such a link, a popup window should appear that allows making the calculation for the one formula that you just clicked.

In this window, there should be a field for editing the formula (in cases where c:geo did not correctly recognize the formula), and at the bottom of the popup there should be your personal cache note (where you define the variable values by writing things like "A=5" etc.).

**Comment by eddiemuc** [comment]
This request is not thought through. If implemented:

* It would no longer allow to add variables for things not directly mentioned in the listing
* It would put too much pressure on the method trying to guess from the listing what are formulas and what not. This function already exists, but it is not (and will never be) perfect
* It would put too much pressure on a free-text formula-parsing function of the personal note, because the parser then has to recognize multiple var-formula-assignments in-between normal text and other stuff like "normal" coordinates. This will get messy
* It would not allow for a place where users can directly see the result of a formula, what is maybe wrong with it and what must be corrected. This place is the var tab

For coordinate calculation, the most common use case for vars, all variables can be accessed and directly edited in the calculated coordinate Dialog already. The variable tab is not needed at all for this use case. It is meant for calculation analysis (e.g. seeing in-between values), calculations beside the pure coordinate calc and for those wanting more comfortable input options e.g. for functions.

Besides, in complex listings the left-right-swipe between description and var tab would just be replaced by up-down-swipes between listing and personal notes

Last but not least, many caches also involve information for variable calculation in waypoints, those are also in another tab. Is the idea to delete this tab as well?

**Comment by dcjkfgdjhd** [comment]
>     * It would no longer allow to add variables for things not directly mentioned in the listing

Yes, the user can simply write the variable directly into the PCN, as they have been doing for years.

>     * It would put too much pressure on the method trying to guess from the listing what are formulas and what not. This function already exists, but it is not (and will never be) perfect

Yes, you have a point there. Of course it is a very rare case that one needs to add formulas that are not at all in the listing, but I guess it should remain possible somehow. Maybe through the menu?

>     * It would put too much pressure on a free-text formula-parsing function of the personal note, because the parser then has to recognize multiple var-formula-assignments in-between normal text and other stuff like "normal" coordinates. This will get messy

Not sure I agree. It only has to recognize simple patterns like "A=5" in the PNC.

>     * It would not allow for a place where users can directly see the result of a formula, what is maybe wrong with it and what must be corrected.

Yes it would. To see that, you would click on the formula link, the popup would pop up, and you would see exactly the same information that you currently see in the variables tab (only for the one formula you clicked, of course).

> For coordinate calculation, the most common use case for vars, all variables can be accessed and directly edited in the calculated coordinate Dialog already. The variable tab is not needed at all for this use case.

Okay, maybe. I have never understood how to use that dialogue. I have tried many times, and I have never succesfully created a new waypoint with correct coordinates *a single time*. Either the dialogue too incomprehensible for me, or it is simply too rigid to be useful. I doubt that there are many people out there who know how to use that feature.

> Besides, in complex listings the left-right-swipe between description and var tab would just be replaced by up-down-swipes between listing and personal notes

No, it would not. Why would you need to swipe up/down? You read through the listing, stage by stage. When you see a formula link, you tap on it. The popup comes up. In the lower part of the popup, the PCN is displayed. Here you can write in your variable values. No need to scroll down at all.

> Last but not least, many caches also involve information for variable calculation in waypoints, those are also in another tab. Is the idea to delete this tab as well?

No, but formulas inside waypoint texts should also become formula links, just like in the listing. Normally, a cache owner will have *either* put all info into to waypoints *or* all of it into the text. So normally for the entire duration of the cache you can stay on the one tab with the info forever, without swiping left or right.

**Comment by eddiemuc** [comment]
Seems to me that almost everything asked for is already implemented except for making recognized formulas clickable (and editable) in listing.

There's already the "scan listing" button in var tab which scans the listing for formulas. Selected formulas can then be edited in the var tab. Transferring this logic into the listing (and make those formulas clickable and first-editable in a popup) is certainly doable. Questions:
* Would this solve the issue at hand?
* Is the performance of the formula recognize algorithm the expected performance for marking formulas in listing?
* How should such clickable formulas in listings be visualized? Underline them?

**Comment by dcjkfgdjhd** [comment]
> Seems to me that almost everything asked for is already implemented except for making recognized formulas clickable (and editable) in listing.

Yes, that was my idea. The functions are already there more or less, they are only not accessible in the ideal way.

>     * Would this solve the issue at hand?

It obviously depends on the exact implementation (and maybe a first implementation could be fine-tuned later), but yes, I think this would already go very far. What is currently so annoying is simply reading the listing, seeing the relevant formula and memorizing it, then swiping left, then searching for the memorized formula to then use it.

If the formula link can be clicked and the popup appears with exactly the same features as there are currently in var tab, then this would solve that inconvenience nicely.

>     * Is the performance of the formula recognize algorithm the expected performance for marking formulas in listing?

The algorithm certainly isn't perfect currently, but often that is not c:geo's fault, but the fault of owners who write formulas in weird formats. But I see this as a separate issue, and I might make separate improvement suggestions for the algorithm in the future. 

But yes, I think the same recognition algorithm can be used for the listing and the var tab.

>     * How should such clickable formulas in listings be visualized? Underline them?

A quick and dirty way would be a fat, blue, underlined link.

But in my mind I imagine a blue background box with rounded corners and bold white text, or something like that. Not sure how easy this would be to do.

**Comment by dcjkfgdjhd** [comment]
Quick and dirty mockup:

![2023-01-08 11 03 02](https://user-images.githubusercontent.com/54892606/211190572-5c132c69-760c-4e7a-b257-4be8bf8b70b6.png)

**Comment by eddiemuc** [comment]
> But in my mind I imagine a blue background box with rounded corners and bold white text, or something like that. 

In principal, everything achievable with HTML is also achievable in listing. Problem is more that this formatting could/might clash with whatever formatting the cache owner has put into his/her listing. In worst case, c:geo overwrites formatting which is somehow relevant to the cache (e.g. to solve a mystery riddle). This formatting should be subtle/restrictive imho.

**Comment by dcjkfgdjhd** [comment]
Yes, I am aware of that. That is why the change I am suggesting is rather minimal.

On the other hand, c:geo already changes the listing formatting so much, that I doubt that a cache where the coloring of the text in the listing matters for solving it, can be solved by looking at it only in c:geo.

**Comment by moving-bits** [comment]
> Currently the "variables" tab is not very useful.
> [...]
> I would completely get rid of the "variables" tab

Please keep in mind that there are very different ways of how people interact with c:geo.
A workflow for user X may be very different from how you work with variables, depending on individual preferences, individual familiarity with c:geo's features and the type of task at hand.

> > For coordinate calculation, the most common use case for vars, all variables can be accessed and directly edited in the calculated coordinate Dialog already. The variable tab is not needed at all for this use case.
> 
> Okay, maybe. I have never understood how to use that dialogue. I have tried many times, and I have never succesfully created a new waypoint with correct coordinates _a single time_. Either the dialogue too incomprehensible for me, or it is simply too rigid to be useful. I doubt that there are many people out there who know how to use that feature.

A typical workflow for me is:
- Copy the formula from within the listing
- Swipe to waypoints tab
- Create a new waypoint with calculated coordinates
- Paste the formula - c:geo then creates all variables automatically for me
- From then on, I work with the variables tab until all variables are set, swiping to waypoints or cache listing on demand, depending on where the info for a variable is listed.

But, as mentioned, this depends on how the cache owner created the listing and how the task is set up.

**Comment by dcjkfgdjhd** [comment]
> * Copy the formula from within the listing
> 
> * Swipe to waypoints tab
> 
> * Create a new waypoint with calculated coordinates

Of course this is a separate issue, and does not conflict with the suggestions discussed here, but in case you are interested:

That is exactly the part that has always confused me: that there is no way to get the formulas directly from within the "calculate coordinates" dialogue, but that you have to manually copy it in.

**Comment by dcjkfgdjhd** [comment]
> Quick and dirty mockup:
> 
> ![2023-01-08 11 03 02](https://user-images.githubusercontent.com/54892606/211190572-5c132c69-760c-4e7a-b257-4be8bf8b70b6.png)

Further idea: As soon as all variables needed to evaluate one of the blue boxes have been set to some values, the formula should no longer be displayed, but the content of the blue box should simply become the result of the calculation.

So step by step, the listing would transform itself, and eventually it would contain all "solved" coordinates.
